keylime/tpm/tpm_main.py

```diff
@@ -40,11 +41,89 @@ def encryptAIK(uuid: str, ek_tpm: bytes, aik_tpm: bytes) -> Optional[Tuple[bytes
             keyblob = base64.b64encode(credentialblob)
 
         except Exception as e:
-            logger.error("Error encrypting AIK: %s", str(e))
+            logger.error("Error encrypting AIK with EK: %s", str(e))
             logger.exception(e)
             raise
 
         return (keyblob, key)
+    
+    @staticmethod
+    def verify_aik_with_iak(uuid: str, aik_tpm: bytes, 
+            iak_tpm: bytes, iak_attest: bytes, iak_sign: bytes) -> bool:
+        
+        # check UUID in certify matches UUID registering
+        if iak_attest[44:(44+len(uuid))] != bytes(uuid, 'utf-8'):
+            logger.warning("Agent %s AIK verification failed, uuid does not match attest info", uuid)
+            return False
+
+        #check aik in certify matches aik being registered
+        if tpm2_objects.get_tpm2b_public_name(aik_tpm) != iak_attest[107:141].hex():
+            logger.warning(" Agent %s AIK verification failed, name of aik does not match attest info", uuid)
+            return False
+
+        # generate digest of attest info
+        attestd, attestpath = tempfile.mkstemp()
+        with open(attestpath, 'wb') as f:
+            f.write(iak_attest)
+        digestd, digestpath = tempfile.mkstemp()
+        command = [
+            "tpm2_hash",
+            "-C",
+            "e",
+            "-g",
+            "sha256",
+            "-o",
+            digestpath,
+            attestpath
+        ]
+        EXIT_SUCCESS: int=0
+        retDict = cmd_exec.run(cmd=command, expectedcode=EXIT_SUCCESS, raiseOnError=False)
+
+        # process iak pub key and import into tpm, get pub key context from tpm
+        iakpub = tpm2_objects.pubkey_from_tpm2b_public(iak_tpm)
```

Comment:
You should call this function here for signature verification:

https://github.com/keylime/keylime/blob/master/keylime/tpm/tpm_util.py#L23

like called from here:

https://github.com/keylime/keylime/blob/master/keylime/tpm/tpm_util.py#L159

The iakpub is the public key that can be used as 1st parameter to verify().

The signature (2nd) parameter is probably what you have as `iak_sign`. It should be 256bytes long for an RSA2048 key.

The you need to calculate the hash over what you want to verify and pass the hash (bytes) as 3rd parameter.

The forth parameter is you hash_func that you can get like this:

https://github.com/keylime/keylime/blob/master/keylime/tpm/tpm_util.py#L148

The hash_alg is the TPM identifier of the hash that you should be able to extract from one of these data  structures here. 0xb represents SHA256.

Reply:
Thanks for the advice, I have removed the hard coded algorithm and removed all use of cli tpm2_tools from the pr in the most recent commit.